pcsx2/DEV9/AdapterUtils.cpp

```diff
@@ -244,30 +247,70 @@ std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 
 	return std::nullopt;
 }
-#elif defined(__POSIX__)
-#ifdef __linux__
+#else
 std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 {
-	struct ifreq ifr;
-	strcpy(ifr.ifr_name, adapter->ifa_name);
+	MAC_Address macAddr = {};
+#if defined(AF_LINK)
+	ifaddrs *adapterInfo;
+	const int error = getifaddrs(&adapterInfo);
+	if (error)
+	{
+		Console.Error("DEV9: Failed to get adapter information %s", error);
+		return std::nullopt;
+	}
 
-	int fd = socket(AF_INET, SOCK_DGRAM, 0);
-	int ret = ioctl(fd, SIOCGIFHWADDR, &ifr);
-	close(fd);
+	const ScopedGuard adapterInfoGuard = [&adapterInfo]() {
+  		freeifaddrs(adapterInfo);
+	};
 
-	if (ret == 0)
-		return *(MAC_Address*)ifr.ifr_hwaddr.sa_data;
+	for (const ifaddrs *po = adapterInfo; po; po = po->ifa_next)
+	{
+		if (strcmp(po->ifa_name, adapter->ifa_name))
+			continue;
 
-	return std::nullopt;
-}
+		if (po->ifa_addr->sa_family != AF_LINK)
+			continue;
+
+		// We have a valid MAC address.
+		std::memcpy(&macAddr, LLADDR(reinterpret_cast<sockaddr_dl*>(po->ifa_addr)), sizeof(macAddr));
+		return macAddr;
+	}
+	freeifaddrs(adapterInfo);
+	Console.Error("DEV9: Failed to get MAC address for adapter using AF_LINK");
 #else
-std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
-{
-	Console.Error("DEV9: Unsupported OS, can't get MAC address");
+	const int sd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP);
+	if (sd < 0)
+	{
+		Console.Error("DEV9: Failed to open socket");
+		return std::nullopt;
+	}
+	struct ifreq ifr = {};
```

Comment:
Could use a scope guard here as well, to avoid the `close(sd)` on all paths

```cpp
  const ScopedGuard sd_guard = [&sd]() {
    close(sd);
  });
```

Reply:
Done!